proposals/0123-block-revenue-distribution.md

```diff
@@ -0,0 +1,174 @@
+the core protocol. For example, the Cogent validator diverts part of its fee
+revenue to NFT holders. But it's up the NFT holders to audit and hold Cogent
+accountable to a specific commission rate.
+
+Another alternative is Jito's mechanism for block "tips" (not fees, but the idea
+is similar). Jito's validator implementation includes a tip distribution program
+which it instructs validator operators to divert all of their tips to but cannot
+enforce perfect compliance. It's up to stakers and the Jito team to audit
+compliance by validator operators. This mechanism requires trusting a
+third-party (in this case Jito) to calculate reward distribution in an accurate
+and fair manner. It also relies on using a merkle tree to distribute fees to all
+stake accounts and the distributed fees are not automatically staked in
+recipient stake accounts.
+
+## New Terminology
+
+NA
+
+## Detailed Design
+
+### Block Revenue Collection
+
+After all transactions are processed in a block for a given leader, rather than
+collecting all block revenue into the validator identity account, the protocol
+will check if the validator's vote account has specified a commission rate and
+collector addresses in the new vote account version described in [SIMD-0185].
+In order to eliminate the overhead of tracking the latest fee collector address
+and commission of each vote account, the vote account state at the beginning of
+the previous epoch MUST be used. This is the same vote account state used to
+build the leader schedule for the current epoch.
+
+If the commission rate and collector account aren't set, all revenue will be
+collected into the validator's identity account as before. If the commission
+rate and collector account *are* specified, the calculated commission MUST be
+deposited in the collector account and the remaining rewards MUST be deposited
+into the validator's vote account for later distribution to stakers.
+
+[SIMD-0185]: https://github.com/solana-foundation/solana-improvement-documents/pull/185
+
+#### Commission Collection
+
+The commission amount MUST be calculated by first multiplying the rate by the
+amount of revenue and then using integer division to divide by `10,000`. If the
+commission amount is non-zero, the fee collector account MUST be loaded and
+checked for the following conditions:
+
+1. account is system program owned AND
+2. account is rent-exempt after depositing the commission.
+
+If the conditions are met, the commission amount MUST be deposited into the fee
+collector account. If either of these conditions is violated, the commission
+amount MUST be burned.
+
+#### Delegator Rewards Collection
+
+The delegator rewards amount MUST be calculated by subtracting the calculated
+commission from the block fee revenue. If the delegator rewards amount is
+non-zero, it must be added to the vote state field `pending_delegator_rewards`
+and added to the balance of vote account.
```

Comment:
This won't work currently because the vote account could have been closed by the time the validator produces their blocks. Vote accounts are close-able as long as they haven't earned any vote credits in both the current epoch or previous epoch. We might need to amend the vote program to prevent closing vote accounts that are part of the leader schedule.

Reply:
I think the best way to handle this edge case is to burn any delegator rewards that can't be recorded in an active vote account. Delegators that observe that their delegated vote account hasn't been voting for an epoch have time to redelegate and should do so because they will be missing out on inflation rewards anyways.

Reply:
> We might need to amend the vote program to prevent closing vote accounts that are part of the leader schedule.

i think there are other ways we can investigate to synchronize vote account liveness with the leader schedule

> I think the best way to handle this edge case is to burn any delegator rewards that can't be recorded in an active vote account. Delegators that observe that their delegated vote account hasn't been voting for an epoch have time to redelegate and should do so because they will be missing out on inflation rewards anyways.

burn seems fine for now. inattentive stake is nothing to be rewarded

Reply:
Addressed in https://github.com/solana-foundation/solana-improvement-documents/pull/123/commits/9ae2000191f7556a9609f6cab85b8e24a7b97b50